File: Template_Method_Pattern/tmp.py
from abc import ABC, abstractmethod

class CaffeineBeverage:
    def prepare(self) -> None:
        self.boilWater()
        self.brew()
        self.pourInCup()
        self.addCondiments()

    @abstractmethod
    def brew(self) -> None: pass

    @abstractmethod
    def addCondiments(self) -> None: pass

    def boilWater(self) -> None: print("Boiling water")
    def pourInCup(self) -> None: print("Pouring into cup")


class Coffee(CaffeineBeverage):
    def brew(self) -> None: print("Dipping coffee through filter")
    def addCondiments(self) -> None: print("Adding suger and milk")

    
class Tea(CaffeineBeverage):
    def brew(self) -> None: print("Steeping the tea")
    def addCondiments(self) -> None: print("Adding lemon")


class CaffeineBeverageWithHook(ABC):
    def prepare(self) -> None:
        self.boilWater()
        self.brew()
        self.pourInCup()
        if self.wantCondiments():
            self.addCondiments()

    @abstractmethod
    def brew(self) -> None: pass

    @abstractmethod
    def addCondiments(self) -> None: pass

    @abstractmethod
    def wantCondiments(self) -> bool: return True

    def boilWater(self) -> None: print("Boiling water")
    def pourInCup(self) -> None: print("Pouring into cup")


class CoffeeWithHook(CaffeineBeverageWithHook):
    def brew(self) -> None: print("Dipping coffee through filter")
    def addCondiments(self) -> None: print("Adding suger and milk")

    def wantCondiments(self) -> bool:
        ans: str = input("You want Condoiments (y/n)? ")

        if ans == 'y': return True
        return False
    

def main()->None:
    tea = Tea()
    tea.prepare()

    coffee = Coffee()
    coffee.prepare()

    print("----------------------------------------------------")

    coffeeWithHook = CoffeeWithHook()
    coffeeWithHook.prepare()
    

if __name__ == '__main__':
    main()

File: Template_Method_Pattern/tmp.cpp
#include <algorithm>
#include <iostream>
#include <memory>
#include <vector>

class CaffeineBeverage {
   public:
    const void prepare() {
        this->boilWater();
        this->brew();
        this->pourInCup();
        this->addCondiments();
    }

	protected:
    virtual void brew() = 0;
    virtual void addCondiments() = 0;

    void boilWater() { std::cout << "Boiling water\n"; }
    void pourInCup() { std::cout << "Pouring into cup\n"; }
};

class Coffee : public CaffeineBeverage {
   protected:
    void brew() override { std::cout << "Dipping coffee through filter\n"; }
    void addCondiments() override { std::cout << "Adding suger and milk\n"; }
};

class Tea : public CaffeineBeverage {
   protected:
    void brew() override { std::cout << "Steeping the tea\n"; }
    void addCondiments() override { std::cout << "Adding lemon\n"; }
};

class CaffeineBeverageWithHook {
   public:
    const void prepare() {
        this->boilWater();
        this->brew();
        this->pourInCup();
        if (this->wantCondiments())
            this->addCondiments();
    }

   protected:
    virtual void brew() = 0;
    virtual void addCondiments() = 0;

    void boilWater() { std::cout << "Boiling water\n"; }
    void pourInCup() { std::cout << "Pouring into cup\n"; }

    virtual bool wantCondiments() { return true; }
};

class CoffeeWithHook : public CaffeineBeverageWithHook {
   protected:
    void brew() override { std::cout << "Dipping coffee through filter\n"; }
    void addCondiments() override { std::cout << "Adding suger and milk\n"; }

    bool wantCondiments() override {
        char ans;
        std::cout << "You want condiments? (y/n) ";
        std::cin >> ans;

        if (ans == 'y')
            return true;
        return false;
    }
};

int main() {
    std::unique_ptr<CaffeineBeverage> tea = std::make_unique<Tea>();
    std::unique_ptr<CaffeineBeverage> coffee = std::make_unique<Coffee>();

    tea->prepare();
    coffee->prepare();

    std::cout << "----------------------------------------------------\n";

    std::unique_ptr<CaffeineBeverageWithHook> coffeeWithHook = std::make_unique<CoffeeWithHook>();

    coffeeWithHook->prepare();
}